Code/Source/solver/mat_models.cpp

```diff
@@ -188,116 +360,120 @@ void get_pk2cc(const ComMod& com_mod, const CepMod& cep_mod, const dmnType& lDmn
+      if (ustruct) {
+        std::runtime_error("[get_pk2cc] St.Venant-Kirchhoff material model not valid for ustruct physics.");
+      }
+      
       double g1 = stM.C10;         // lambda
       double g2 = stM.C01 * 2.0;   // 2*mu
 
-      S = g1*trE*Idm + g2*E;
-      CC = g1 * ten_dyad_prod(Idm, Idm, nsd) + g2*ten_ids(nsd);
+      S += g1*trE*Idm + g2*E;
+      CC += g1 * ten_dyad_prod_eigen<nsd>(Idm, Idm) + g2*ten_ids_eigen<nsd>();
     } break;
 
     // modified St.Venant-Kirchhoff
     case ConstitutiveModelType::stIso_mStVK: {
+      if (ustruct) {
+        std::runtime_error("[get_pk2cc] Modified St.Venant-Kirchhoff material model not valid for ustruct physics.");
+      }
+
       double g1 = stM.C10; // kappa
       double g2 = stM.C01;  // mu
 
-      S = g1*log(J)*Ci + g2*(C-Idm);
-      CC = g1 * ( -2.0*log(J)*ten_symm_prod(Ci, Ci, nsd) +
-         ten_dyad_prod(Ci, Ci, nsd) ) + 2.0*g2*ten_ids(nsd);
+      S += g1*log(J)*Ci + g2*(C-Idm);
+      CC += g1 * ( -2.0*log(J)*ten_symm_prod_eigen<nsd>(Ci, Ci) +
+         ten_dyad_prod_eigen<nsd>(Ci, Ci) ) + 2.0*g2*ten_ids_eigen<nsd>();
     } break;
 
     // NeoHookean model
     case ConstitutiveModelType::stIso_nHook: {
-      double g1 = 2.0 * stM.C10;
-      auto Sb = g1*Idm;
 
-      // Fiber reinforcement/active stress
-      Sb += Tfa * mat_dyad_prod(fl.col(0), fl.col(0), nsd);
+      // Compute fictious stress and elasticity tensor
+      EigenMatrix<nsd> S_bar = 2.0 * stM.C10 * Idm;
+      EigenTensor<nsd> CC_bar; CC_bar.setZero();
 
-      double r1 = g1 * Inv1 / nd;
-      for (int j = 0; j < S.ncols(); j++) {
-        for (int i = 0; i < S.nrows(); i++) {
-          S(i,j) = J2d*Sb(i,j) - r1*Ci(i,j);
-        }
-      }
+      // Add fiber reinforcement/active stress
+      S_bar += Tfa * (fl.col(0) * fl.col(0).transpose());
 
-      CC = (-2.0/nd) * ( ten_dyad_prod(Ci, S, nsd) + ten_dyad_prod(S, Ci, nsd));
-      S += p*J*Ci;
-      CC += 2.0*(r1 - p*J) * ten_symm_prod(Ci, Ci, nsd)  +  (pl*J - 2.0*r1/nd) * ten_dyad_prod(Ci, Ci, nsd);
+      // Compute and add isochoric stress and elasticity tensor
+      auto [S_iso, CC_iso] = bar_to_iso<nsd>(S_bar, CC_bar, J2d, C, Ci);
+      S += S_iso;
+      CC += CC_iso;
 
     } break;
 
-    //  Mooney-Rivlin model
+    // Mooney-Rivlin model
     case ConstitutiveModelType::stIso_MR: {
-      double g1  = 2.0 * (stM.C10 + Inv1*stM.C01);
-      double g2  = -2.0 * stM.C01;
-      auto Sb = g1*Idm + g2*J2d*C;
 
-      // Fiber reinforcement/active stress
-      Sb = Sb + Tfa*mat_dyad_prod(fl.col(0), fl.col(0), nsd);
+      // Compute fictious stress and elasticity tensor
+      EigenMatrix<nsd> S_bar = 2.0 * (stM.C10 + Inv1 * stM.C01) * Idm 
+                              -2.0 * stM.C01 * J2d * C;
 
-      g1  = 4.0*J4d* stM.C01;
-      auto CCb = g1 * (ten_dyad_prod(Idm, Idm, nsd) - ten_ids(nsd));
+      EigenTensor<nsd> CC_bar = 4.0 * J4d * stM.C01 * (ten_dyad_prod_eigen<nsd>(Idm, Idm) - ten_ids_eigen<nsd>());
```

Comment:
Note that you don't need to use `<nsd>` for template functions that use `nsd` in its parameters
```
EigenTensor<nsd> CC_bar = 4.0 * J4d * stM.C01 * (ten_dyad_prod_eigen(Idm, Idm) - ten_ids_eigen<nsd>());
```

Reply:
Ah interesting!